### PR TITLE
cmd/evm: validate blockchain tests poststate account storage

### PR DESCRIPTION
**Pull request by danceratopz**
This PR additionally verifies the accounts' storage as specified in a blockchain test's `postState` field (as described in the "Expect" section [here](https://ethereum-tests.readthedocs.io/en/latest/test_filler/blockchain_filler.html#expect)).

### CODE REVIEW

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM, weird that we didn't verify it before :D

File: tests/block_test_util.go [hunk 1]
```diff
@@ -330,6 +330,12 @@ func (t *BlockTest) validatePostState(statedb *state.StateDB) error {
 		if nonce2 != acct.Nonce {
 			return fmt.Errorf("account nonce mismatch for addr: %s want: %d have: %d", addr, acct.Nonce, nonce2)
 		}
+		for k, v := range acct.Storage {
+			v2 := statedb.GetState(addr, k)
+			if v2 != v {
+				return fmt.Errorf("account storage mismatch for addr: %s, slot: %x, want: %x, have: %x", addr, k, v, v2)
+			}
+		}
```
Comment (karalabe) [inline tests/block_test_util.go:338]: We should probably also add acct.RootHash?

**Comment by holiman** [comment]
Let me just outline how I think this works, and let me know if you disagree with this: 

So, this expect-section, it does really only check that the _test_ works. It's meant for the test-author to verify that "If the test does what it's supposed to, then the nonce of `X` should be `2`, and the slot `Y` at `Z` should be `0x123`. 

- This expect-section is not exhaustive (not full post-state)
- It is also not auto-generated, but put there manually by the author. 

---

The actual _test_ to check "does client X" pass the test, doesn't actually use the info in this section: all it needs to do is verify the stateroot, which is both exhaustive (contains _all_ state, not just snippets) and auto-generated when the test was filled. 


---

If this understanding is correct, then I don't think we need to also check the `acct.RootHash` as @karalabe suggests, since it's not relevant for that section.

File: tests/block_test_util.go [hunk 2]
```diff
@@ -330,6 +330,12 @@ func (t *BlockTest) validatePostState(statedb *state.StateDB) error {
 		if nonce2 != acct.Nonce {
 			return fmt.Errorf("account nonce mismatch for addr: %s want: %d have: %d", addr, acct.Nonce, nonce2)
 		}
+		for k, v := range acct.Storage {
+			v2 := statedb.GetState(addr, k)
+			if v2 != v {
+				return fmt.Errorf("account storage mismatch for addr: %s, slot: %x, want: %x, have: %x", addr, k, v, v2)
```
Comment (holiman) [inline tests/block_test_util.go:336]:
```suggestion
			
			if have := statedb.GetState(addr, k); have != v {
				return fmt.Errorf("account storage mismatch for addr: %s, slot: %x, want: %x, have: %x", addr, k, v, have)
```

**Review by holiman — APPROVED** [review]
Minor nit, otherwise lgtm

File: tests/block_test_util.go (hunk 1 above)
Comment (danceratopz) [reply tests/block_test_util.go:338]: I agree with @holiman's comment [below](https://github.com/ethereum/go-ethereum/pull/28443#issuecomment-1787391595), this is tested via the `stateRoot` specified in the final block in the `blocks` section of the test fixture.

**Comment by danceratopz** [comment]
> Let me just outline how I think this works, and let me know if you disagree with this:
> 
> So, this expect-section, it does really only check that the _test_ works. It's meant for the test-author to verify that "If the test does what it's supposed to, then the nonce of `X` should be `2`, and the slot `Y` at `Z` should be `0x123`.
> 
>     * This expect-section is not exhaustive (not full post-state)
> 
>     * It is also not auto-generated, but put there manually by the author.
> 
> 
> The actual _test_ to check "does client X" pass the test, doesn't actually use the info in this section: all it needs to do is verify the stateroot, which is both exhaustive (contains _all_ state, not just snippets) and auto-generated when the test was filled.
> 
> If this understanding is correct, then I don't think we need to also check the `acct.RootHash` as @karalabe suggests, since it's not relevant for that section.

Hey @holiman, thanks for clarifying, yes, I do agree with this. This information is extremely helpful during test implementation, but it could even be omitted from the test fixture that gets executed against clients. Although it does help add context to the JSON fixture (without the need of the corresponding test source). Since it does get included, it seems reasonable to check it's correct.

Prior to pushing, I re-ran `blocktest` with the storage check against all `execution-spec-test` and `ethereum/tests/BlockchainTests` to verify that no fails were introduced. I realized afterwards that it would be very strange if there was a fail, indeed if there was, it would be an error in the fixture's `postState` section (^="expect" in the docs), and not the evm.

File: tests/block_test_util.go (hunk 2 above)
Comment (danceratopz) [reply tests/block_test_util.go:336]: I went with `v2` to match the style of the surrounding code, but I also prefer `have`. What do you think about renaming the other variables, e.g., `code2` -> `have_code`, for consistency?

**Pull request merged**